### ISSUE TITLE
Broken pip in Kadalu containers

### ISSUE THREAD

**Issue by leelavg** [kadalu/kadalu#616]
## Issue:
1. Neither in amd64 or arm images `pip` is in usable state except in `builder` image
2. This is a pain when we want to check pkgs installed in the container (for ex: https://github.com/kadalu/kadalu/issues/611)

## RC:
1. Pip depends on `disutils.core` and `appdirs` big time, we are installing directly/indirectly these via https://github.com/kadalu/kadalu/blob/61589acec98c1b2e91c0cf4fd77473dc8c2e5987/extras/Dockerfile.builder#L14-L15
2. Even though we state venv to use site-packages https://github.com/kadalu/kadalu/blob/61589acec98c1b2e91c0cf4fd77473dc8c2e5987/extras/Dockerfile.builder#L24 these (distutils/system wide installed pkgs etc) aren't availble in other component (operator, server, csi) images
3. Some pkgs we can directly copy and make them work like https://github.com/kadalu/kadalu/blob/61589acec98c1b2e91c0cf4fd77473dc8c2e5987/csi/Dockerfile#L28 but it's not working for pip dependent pkgs
4. The way ubuntu image sees python is, `apt` will not install whole python core embargo when we install `python3` it leaves a lot of system pkgs (like venv, distutils etc)

## Possible Fixes (we can pick anyone from below):
With https://github.com/kadalu/kadalu/pull/592 we are in a good state in controlling all third-party dependencies and I want to push it further to take control of system pkgs as well
1. Install `python3-distutils` in component images and pip will work in amd64 images and workaround `appdirs` dependency in arm images to make pip work
2. Bring back python only multistage image and hack around it to use this single image as base wrt all python system and third-party pkgs
3. Swap ubuntu with python-3.8.11-slim-buster for all component images (ig it's totally compatible with `builder`)

Any comments/corrections are more than welcome, btw I vote for fix 3 😉

**Comment by amarts** [comment]
We surely can experiment with point 3, as it looks simple enough. For testing we can push to arm branch, so it can be tested first.

**Comment by leelavg** [comment]
Ack, pls rebase one of the arm branches on devel and I'll send the PR based on that.

Since, we aren't associating `arm` tags anymore I think it'll be a fast-forward merge 🤔

**Comment by aravindavk** [comment]
How about copying python packages from the builder image to `/kadalu` and set this path to `PYTHON_PATH` env variable. No `pip` or `venv` required in the final image.

**Comment by leelavg** [comment]
That would also work to get away with pip & venv.

However, it'll make listing installed packages pretty hard so more debugging efforts are needed.

Dependency tree will be like, pip -> distutils.core (platform dependent)

Nevertheless, we can give a try making `/kadalu` as `PYTHON_PATH`

**Comment by leelavg** [comment]
With below diff, things are looking good but image sizes shot up >3x, need to work on one of below:
1. Reduce size with python:3.8.11-slim-buster (maybe symlink to `/kadalu/bin/python3` directly and whole layer will not be copied 🤔)
2. Fix existing (ubuntu) image not to fail with import errors and get away without using pip
3. Fix existing (ubunut) image and still retain usage of pip

``` diff
diff --git a/csi/Dockerfile b/csi/Dockerfile
index 8cc21bc..bb8ba36 100644
--- a/csi/Dockerfile
+++ b/csi/Dockerfile
@@ -8,11 +8,12 @@ COPY csi-requirements.txt /tmp/
 
 RUN python3 -m pip install -r /tmp/csi-requirements.txt
 
-FROM ubuntu:20.04 as prod
+FROM python:3.8.11-slim-buster as prod
 ENV DEBIAN_FRONTEND=noninteractive
 RUN apt-get update -yq && \
-    apt-get install -y --no-install-recommends attr python3 libtirpc3 bash inotify-tools ssh && \
+    apt-get install -y --no-install-recommends attr libtirpc3 bash inotify-tools ssh && \
     apt-get -y clean && \
+    ln -sf /usr/local/bin/python3 /usr/bin/python3 && \
     rm -rf /var/lib/apt/lists/*
 
 ENV PATH="/kadalu/bin:/opt/bin:/opt/sbin:$PATH"
diff --git a/operator/Dockerfile b/operator/Dockerfile
index 835e562..c1a94d8 100644
--- a/operator/Dockerfile
+++ b/operator/Dockerfile
@@ -8,12 +8,12 @@ COPY operator-requirements.txt /tmp/
 
 RUN python3 -m pip install -r /tmp/operator-requirements.txt
 
-FROM ubuntu:20.04 as prod
+FROM python:3.8.11-slim-buster as prod
 
 ENV DEBIAN_FRONTEND=noninteractive
 RUN apt-get update -yq && \
-    apt-get install -y --no-install-recommends python3 && \
     apt-get -y clean && \
+    ln -sf /usr/local/bin/python3 /usr/bin/python3 && \
     rm -rf /var/lib/apt/lists/*
 
 ENV PATH="/kadalu/bin:$PATH"
diff --git a/server/Dockerfile b/server/Dockerfile
index b7c6fdb..f29cab5 100644
--- a/server/Dockerfile
+++ b/server/Dockerfile
@@ -8,7 +8,7 @@ COPY server-requirements.txt /tmp/
 
 RUN python3 -m pip install -r /tmp/server-requirements.txt
 
-FROM ubuntu:20.04 as prod
+FROM python:3.8.11-slim-buster as prod
 
 ENV DEBIAN_FRONTEND=noninteractive
 ENV PATH="/kadalu/bin:/opt/bin:/opt/sbin:$PATH"
@@ -19,8 +19,9 @@ COPY --from=builder /opt /opt
 COPY --from=builder /kadalu /kadalu
 
 RUN apt-get update -yq && \
-    apt-get install -y --no-install-recommends attr python3 xfsprogs libtirpc3 sqlite3 && \
+    apt-get install -y --no-install-recommends attr xfsprogs libtirpc3 sqlite3 && \
     apt-get -y clean && \
+    ln -sf /usr/local/bin/python3 /usr/bin/python3 && \
     rm -rf /var/lib/apt/lists/*
 
 # copy pyxattr installed in pypi_source image
```

```
-> docker images | grep kadalu-
kadalu/kadalu-operator                                devel                4ff9b9f3e54b   8 minutes ago       215MB
kadalu/kadalu-csi                                     devel                ac52169463ca   8 minutes ago       211MB
kadalu/kadalu-server                                  devel                0ae59e4d2dab   9 minutes ago       249MB
```

Edit: My mistake, I was comparing compressed sizes to extracted sizes, it'll be almost  same even after using `python:slim` as component base.

**Comment by leelavg** [comment]
Fixed by #620
